deployments/advanced-analysis/services/advanced-analysis/src/main.py

```diff
@@ -0,0 +1,379 @@
+from contextlib import asynccontextmanager
+from datetime import datetime
+import os
+import time
+from typing import Any
+import uuid
+from zoneinfo import ZoneInfo
+
+from fastapi import FastAPI, Request, Response, Depends, status, HTTPException, Query
+from fastapi.middleware.cors import CORSMiddleware
+from fastapi.responses import StreamingResponse
+from starlette_context import context, request_cycle_context
+from stream_zip import stream_zip
+import structlog
+from structlog.contextvars import bind_contextvars, clear_contextvars
+from uvicorn.protocols.utils import get_path_with_query_string
+
+from auth import ProtectedAny, ScopeTags
+from jobs import (
+    check_customer_advanced_analysis_usage,
+    create_advanced_analysis_job,
+    delete_advanced_analyses,
+    get_advanced_analyses_for_admin,
+    get_advanced_analyses_for_base_user,
+    get_advanced_analyses_download_info_for_base_user,
+    get_advanced_analyses_download_info_for_admin,
+)
+from utils.db import AsyncpgPoolDep
+from utils.logging import setup_logger, bind_context_to_logger
+from utils.s3 import yield_s3_objects, generate_presigned_url
+from core.config import DATABASE_URL, DASHBOARD_URL, PULSE3D_UPLOADS_BUCKET
+from models.models import (
+    GetAdvancedAnalysesResponse,
+    GetAdvancedAnalysisUsageResponse,
+    PostAdvancedAnalysesRequest,
+    PostAdvancedAnalysesDownloadRequest,
+)
+
+setup_logger()
+logger = structlog.stdlib.get_logger("api.access")
+
+asyncpg_pool = AsyncpgPoolDep(dsn=DATABASE_URL)
+
+
+@asynccontextmanager
+async def lifespan(app: FastAPI):
+    await asyncpg_pool()
+    yield
+
+
+app = FastAPI(openapi_url=None, lifespan=lifespan)
+
+app.add_middleware(
+    CORSMiddleware,
+    allow_origins=[DASHBOARD_URL],
+    allow_credentials=True,
+    allow_methods=["*"],
+    allow_headers=["*"],
+)
+
+
+@app.middleware("http")
+async def middleware(request: Request, call_next) -> Response:
+    request.state.pgpool = await asyncpg_pool()
+    # clear previous request variables
+    clear_contextvars()
+    # get request details for logging
+    if (client_ip := request.headers.get("X-Forwarded-For")) is None:
+        client_ip = f"{request.client.host}:{request.client.port}"
+
+    url = get_path_with_query_string(request.scope)
+    http_method = request.method
+    http_version = request.scope["http_version"]
+    start_time = time.perf_counter_ns()
+
+    bind_contextvars(url=str(request.url), method=http_method, client_ip=client_ip)
+
+    with request_cycle_context({}):
+        response = await call_next(request)
+
+        process_time = time.perf_counter_ns() - start_time
+        status_code = response.status_code
+
+        logger.info(
+            f"""{client_ip} - "{http_method} {url} HTTP/{http_version}" {status_code}""",
+            status_code=status_code,
+            duration=process_time / 10**9,
+            **context,
+        )
+
+    return response
+
+
+# ROUTES
+
+
+@app.get("/versions")
```

Comment:
when we discussed adding new notifications-related endpoints, we mentioned we should make it more modular so it could be moved, tested, etc.  I've been considering using a service layer (see e.g. [this post](https://medium.com/@kacperwlodarczyk/fast-api-repository-pattern-and-service-layer-dad43354f07a)), but that's something you may want to punt on for now

Reply:
That would definitely be a good idea long term. It might be worth putting together a design doc first for how we want split everything up so that we're all on the same page as to how we'd like to do it